dagstore.go

```diff
@@ -0,0 +1,602 @@
+				_ = d.queueTask(&Task{Op: OpShardAcquire, Shard: tsk.Shard, Resp: acqCh}, true)
+			}
+			s.wAcquire = s.wAcquire[:0]
+
+		case OpShardAcquire:
+			if s.state != ShardStateAvailable && s.state != ShardStateServing {
+				// shard state isn't active yet; make this acquirer wait.
+				s.wAcquire = append(s.wAcquire, tsk.Resp)
+				break
+			}
+
+			s.state = ShardStateServing
+			s.refs++
+			go func(ctx context.Context, tsk *Task, k shard.Key, mnt mount.Mount) {
+				reader, err := mnt.Fetch(ctx)
+				if err != nil {
+					err = fmt.Errorf("failed to acquire reader of mount: %w", err)
+					_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+					tsk.Resp <- ShardResult{Key: s.key, Error: err}
+					return
+				}
+
+				idx, err := d.indices.GetFullIndex(k)
+				if err != nil {
+					err = fmt.Errorf("failed to recover index for shard %s: %w", k, err)
+					_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+					tsk.Resp <- ShardResult{Key: s.key, Error: err}
+					return
+				}
+
+				sa, err := NewShardAccessor(s.key, reader, idx)
+				tsk.Resp <- ShardResult{Key: s.key, Accessor: sa, Error: err}
+			}(d.ctx, tsk, s.key, s.mount)
+
+		case OpShardRelease:
+			if (s.state != ShardStateServing && s.state != ShardStateErrored) || s.refs <= 0 {
+				log.Warnf("ignored illegal request to release shard")
+				break
+			}
+			s.refs--
+
+		case OpShardFail:
+			s.state = ShardStateErrored
+			s.err = tsk.Error
+
+			// can't block the event loop, so launch a goroutine to notify.
+			if s.wRegister != nil {
+				res := ShardResult{
+					Key:   s.key,
+					Error: fmt.Errorf("failed to register shard: %w", tsk.Error),
+				}
+				go func(ch chan ShardResult) { ch <- res }(s.wRegister)
+			}
+
+			// fail waiting acquirers.
+			// can't block the event loop, so launch a goroutine per acquirer.
+			if len(s.wAcquire) > 0 {
+				res := ShardResult{
+					Key:   s.key,
+					Error: fmt.Errorf("failed to acquire shard: %w", tsk.Error),
+				}
+				for _, acqCh := range s.wAcquire {
+					go func(ch chan ShardResult) { ch <- res }(acqCh)
+				}
+				s.wAcquire = s.wAcquire[:0] // empty acquirers.
+			}
+
+			// TODO trigger retries?
+
+		case OpShardDestroy:
+			if s.state == ShardStateServing || s.refs > 0 {
+				res := ShardResult{
+					Key:   s.key,
+					Error: fmt.Errorf("failed to destroy shard; active references: %d", s.refs),
+				}
+				go func(ch chan ShardResult) { ch <- res }(tsk.Resp)
+				break
+			}
+
+			d.lk.Lock()
+			delete(d.shards, s.key)
+			d.lk.Unlock()
+			// TODO are we guaranteed that there are no queued items for this shard?
+		}
+
+		s.Unlock()
+	}
+}
+
+func (d *DAGStore) Close() error {
+	d.cancelFn()
+	d.wg.Wait()
+	return nil
+}
+
+func (d *DAGStore) queueTask(tsk *Task, overflow bool) error {
+	select {
+	case <-d.ctx.Done():
+		return fmt.Errorf("dag store closed")
+	case d.taskCh <- tsk:
+		return nil
+	default:
+	}
+	if overflow {
```

Comment:
I think it would simplify the implementation to use a single queue instead of having a separate taskCh channel and overflow queue

Reply:
I have a better model! This can be simplified by using two chans:

- internal, with buffer 1, used for internal tasks.
- external, with buffer N, used for external tasks.

Reply:
Solved here: https://github.com/filecoin-project/dagstore/pull/22